packages/flutter/test/widgets/text_formatter_test.dart

```diff
@@ -108,5 +108,141 @@ void main() {
         ),
       ));
     });
+
+    test('test length limiting formatter', () {
+      final TextEditingValue actualValue =
+      new LengthLimitingTextInputFormatter(6)
+          .formatEditUpdate(testOldValue, testNewValue);
+
+      // Expecting
+      // a1b(2c3)
+      expect(actualValue, const TextEditingValue(
+        text: 'a1b2c3',
+        selection: const TextSelection(
+          baseOffset: 3,
+          extentOffset: 6,
+        ),
+      ));
+    });
+
+    test('test length limiting formatter with zero-length string', () {
+      testNewValue = const TextEditingValue(
+        text: '',
+        selection: const TextSelection(
+          baseOffset: 0,
+          extentOffset: 0,
+        ),
+      );
+
+      final TextEditingValue actualValue =
+      new LengthLimitingTextInputFormatter(1)
+        .formatEditUpdate(testOldValue, testNewValue);
+
+      // Expecting the empty string.
+      expect(actualValue, const TextEditingValue(
+        text: '',
+        selection: const TextSelection(
+          baseOffset: 0,
+          extentOffset: 0,
+        ),
+      ));
+    });
+
+    test('test length limiting formatter with non-ASCII runes', () {
+      testNewValue = const TextEditingValue(
+        text: '\u{1f984}\u{1f984}\u{1f984}\u{1f984}', // Unicode U+1f984 (UNICORN FACE)
+        selection: const TextSelection(
+          baseOffset: 4,
+          extentOffset: 4,
+        ),
+      );
+
+      final TextEditingValue actualValue =
+      new LengthLimitingTextInputFormatter(2)
+        .formatEditUpdate(testOldValue, testNewValue);
+
+      // Expecting two runes.
+      expect(actualValue, const TextEditingValue(
+        text: '\u{1f984}\u{1f984}',
+        selection: const TextSelection(
+          baseOffset: 2,
+          extentOffset: 2,
+        ),
+      ));
+    });
+
+
+    test('test length limiting formatter with complex Unicode characters', () {
+      // TODO(gspencer): Test additional strings.  We can do this once the
+      // formatter supports Unicode grapheme clusters.
+      //
+      // A formatter with max length 1 should accept:
+      //  - The '\u{1F3F3}\u{FE0F}\u{200D}\u{1F308}' sequence (flag followed by
+      //    a variation selector, a zero-width joiner, and a rainbow to make a rainbow
+      //    flag).
+      //  - The sequence '\u{0058}\u{0346}\u{0361}\u{035E}\u{032A}\u{031C}\u{0333}\u{0326}\u{031D}\u{0332}'
+      //    (Latin X with many composed characters).
+      //
+      // A formatter should not count as a character:
+      //   * The '\u{0000}\u{FEFF}' sequence. (NULL followed by zero-width no-break space).
+      //
+      // A formatter with max length 1 should truncate this to one character:
+      //   * The '\u{1F3F3}\u{FE0F}\u{1F308}' sequence (flag with ignored variation
+      //     selector followed by rainbow, should truncate to just flag).
```

Comment:
it should specifically not truncate it
It should truncate `'\u{1F3F3}\u{FE0F}\u{1F308}X' to a rainbow flag, dropping only the `'X'`.

Reply:
OK, but from what I can tell a rainbow flag is four scalar values: `'\u{1F3F3}\u{FE0F}\u{200D}\u{1F308}'`, not just the three.

[Not that it's definitive, but Chrome renders the one with only three scalar values as two separate characters, flag and rainbow.]

Reply:
OK, spoke with Ian, and we decided that my example is correct: the first two values indicate that we want the "Emoji Flag" (as opposed to just a flag), and it's followed by a rainbow.